csrc/aes_xts.cpp

```diff
@@ -0,0 +1,145 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+#include "buffer.h"
+#include "env.h"
+#include <openssl/evp.h>
+#include <jni.h>
+
+// Tweak and key are passed in the same buffer. The first 16 bytes
+// of this buffer is the tweak and the rest is the key.
+static int const AES_XTS_KEY_INDEX_START = 16;
+
+namespace AmazonCorrettoCryptoProvider {
+
+class AesXtsCipher {
+public:
+    AesXtsCipher(bool for_encryption, unsigned char const* key, unsigned char const* tweak)
+    {
+        ctx_ = EVP_CIPHER_CTX_new();
+        if (ctx_ == nullptr) {
+            throw_openssl(EX_RUNTIME_CRYPTO, "EVP_CIPHER_CTX_new failed.");
+        }
+
+        if (for_encryption) {
+            if (EVP_EncryptInit_ex(ctx_, EVP_aes_256_xts(), nullptr, key, tweak) != 1) {
+                throw_openssl(EX_RUNTIME_CRYPTO, "EVP_EncryptInit_ex failed.");
+            }
+        } else {
+            if (EVP_DecryptInit_ex(ctx_, EVP_aes_256_xts(), nullptr, key, tweak) != 1) {
+                throw_openssl(EX_RUNTIME_CRYPTO, "EVP_DecryptInit_ex failed.");
+            }
+        }
+
+        // this method always returns 1
+        EVP_CIPHER_CTX_set_padding(ctx_, 0);
+    }
+
+    ~AesXtsCipher() { EVP_CIPHER_CTX_free(ctx_); }
+
+    void encrypt(unsigned char* input, int input_len, unsigned char* output)
+    {
+        int out_len = 0;
+        if (EVP_EncryptUpdate(ctx_, output, &out_len, input, input_len) != 1) {
+            throw_openssl(EX_RUNTIME_CRYPTO, "EVP_EncryptUpdate failed.");
+        }
+
+        if (EVP_EncryptFinal_ex(ctx_, output + out_len, &out_len) != 1) {
+            throw_openssl(EX_RUNTIME_CRYPTO, "EVP_EncryptFinal_ex failed.");
+        }
+    }
+
+    void decrypt(unsigned char* input, int input_len, unsigned char* output)
+    {
+        int out_len = 0;
+        if (EVP_DecryptUpdate(ctx_, output, &out_len, input, input_len) != 1) {
+            throw_openssl(EX_RUNTIME_CRYPTO, "EVP_DecryptUpdate failed.");
+        }
+
+        if (EVP_DecryptFinal_ex(ctx_, output + out_len, &out_len) != 1) {
+            throw_openssl(EX_RUNTIME_CRYPTO, "EVP_DecryptFinal_ex failed.");
+        }
+    }
+
+private:
+    EVP_CIPHER_CTX* ctx_;
+};
+
+}
+
+using namespace AmazonCorrettoCryptoProvider;
+
+extern "C" JNIEXPORT void JNICALL Java_com_amazon_corretto_crypto_provider_AesXtsSpi_enc(JNIEnv* env,
+    jclass,
+    jbyteArray jtweakKey,
```

Comment:
Do we need to pack both keys into 1 jbyteArray here? Why not have two separate jbyteArrays? Is this just to avoid an extra `JByteArrayCritical` call?

Also, naming nit, if this byte array contains two different keys, you should follow the `packedTweakKey` naming convention that was used Java-side here as well.

Reply:
It's just to avoid creating one extra `JByteArrayCritical`.
Fixed the naming.